Consensus diagrams

mDuo13 opened the pull request:
- Revises the diagrams uses in the "Consensus" document to match the style for the recent "Intro to Consensus" article.
- Revises the Consensus document itself to better use the diagrams and to fix some outdated style from the previous article.
- Adds a "Transaction Queue" article with a diagram of how the queue interacts with consensus. This article also takes over some of the queue-related content that previously existed in the Transaction Cost page.

<s>This PR currently contains an unused HTML version of the queue+consensus diagram and some images only used in various versions of that diagram. (The initial thought was that making the diagram HTML would be more accessible and flexible, but in practice, it didn't fit well on screen at almost any screen width. I concluded that a simple image with the text reproduced in the document was more flexible and would display better on a variety of viewing devices.) **Reviewers: I'd appreciate your feedback on whether the unused diagram is useful scrap material or should be removed.**</s>

Edit: I have removed the unused HTML diagram and its related images.

mDuo13 committed:
Start revising consensus article diagrams

mDuo13 committed:
Consensus: more revised diagrams

mDuo13 committed:
Consensus: update 'calculate validation' diagram

mDuo13 committed:
Consensus diagrams: new validate diagram; queue page and diagram

mDuo13 committed:
Queue, Consensus: separate out queue mechanics, clarify how ledgers are calculated

mDuo13 committed:
Make Queue+Consensus diagram 2 columns

Needs more work to display better on a variety of screen resolutions

mDuo13 committed:
TxQ diagram: back to single image

mDuo13 commented:
Preview of these docs staged here so you can see the diagrams in context:
_(Note: Ripple VPN is required to view staging.)_
https://developers-staging.ripple.com/consensus.html
https://developers-staging.ripple.com/transaction-queue.html

EDIT: Updated with changes through 1177df2.

ximinez commented on content/concepts/consensus-network/consensus.md:
```diff
@@ -6,35 +6,36 @@ This article provides a high level overview of the XRP Ledger, the information i
+A ledger version has two identifiers. One identifier is its _ledger index_, also called a _sequence number_. Ledger versions are numbered incrementally. For example, if the current ledger version has ledger index of 100, the previous has ledger index 99 and the next has ledger index 101. The other identifier is a _ledger hash_, which is a digital fingerprint of the ledger's contents.
 
-User level changes to the ledger are the results of transactions. Examples of [transactions](transaction-formats.html) include payments, changes to account settings or trust lines, and offers to trade. Each transaction authorizes changes to the ledger and is cryptographically signed by an account owner. Transactions are the only way to authorize changes to an account.
+As servers propose transactions to apply to the ledger, they may create several candidate ledger versions with slightly different contents. These candidate ledger versions have the same ledger index but different ledger hashes. Of the many candidates, only one can become validated. All the other candidate ledger versions are discarded. Thus, there is exactly exactly one validated ledger hash for  ...
```
Duplicated word: `exactly exactly`

ximinez commented on content/concepts/consensus-network/consensus.md:
```diff
@@ -6,35 +6,36 @@ This article provides a high level overview of the XRP Ledger, the information i
+As servers propose transactions to apply to the ledger, they may create several candidate ledger versions with slightly different contents. These candidate ledger versions have the same ledger index but different ledger hashes. Of the many candidates, only one can become validated. All the other candidate ledger versions are discarded. Thus, there is exactly exactly one validated ledger hash for  ...
 
-A ledger instance also contains a set of transactions and metadata about those transactions. The transactions are those that have been applied to the prior ledger to create the new instance. The metadata records precisely the effects of the transaction on the ledger.
+User level changes to the ledger are the results of transactions. Examples of [transactions](transaction-formats.html) include payments, changes to account settings or trust lines, and offers to trade. Each transaction authorizes changes to the ledger and is cryptographically signed by an account owner. Transactions are the only way to authorize changes to an account, or to change anything else i ...
```
This reads a little awkwardly to me: `Each transaction authorizes changes to the ledger`. I'm having some noun/verb confusion with "changes". I think it may be clearer as `Each transaction authorizes one or more changes to the ledger`

ximinez commented on content/concepts/consensus-network/consensus.md:
```diff
@@ -54,61 +55,82 @@ Important: Some [`rippled` APIs](rippled-api.html) provide provisional results,
+
+_Figure 8: Ledger is Validated When Supermajority of Peers Calculate the Same Result — Servers compare their calculated ledger with the hashes received from chosen validators. If not in agreement, the server must re-calculate or retrieve the correct ledger._
+
+Servers in the network recognize a ledger instance as validated when a supermajority of the peers have signed and broadcast the same validation hash <a href="#footnote_8" id="footnote_from_8"><sup>8</sup></a>. Going forward, transactions are applied to this updated and now validated ledger with sequence number N+1.
```
Should `supermajority of the peers` instead say `supermajority of chosen validators` in this and the last couple of paragraphs? Because when I see "peers", I think "the nodes I'm connected to", and I don't care about their validations (unless they happen to be a chosen validator).

ximinez commented on content/concepts/consensus-network/consensus.md:
```diff
@@ -54,61 +55,82 @@ Important: Some [`rippled` APIs](rippled-api.html) provide provisional results,
+3. Process each transaction according to its instructions, in order. Update the ledger's state data accordingly.
 
-_Figure 8: Ledger is Validated When Supermajority of Peers Calculate the Same Result — Nodes compare their calculated ledger with the hashes received from chosen validators. If not in agreement, the node must re-calculate or retrieve the correct ledger._
+    If the transaction cannot be successfully executed, destroy the XRP transaction cost only. Include the transaction with a [`tec`-class result code](tec-codes.html).
```
`tec` results also cause the account's sequence number to increment.

mDuo13 replied:
Linked to a (rewritten) footnote on tec codes and modified the other case that said something similar.

ximinez commented on content/concepts/consensus-network/consensus.md:
```diff
@@ -150,24 +172,22 @@ Best practices for applications submitting transactions include:
 
-<a href="#footnote_from_1" id="footnote_1"><sup>1</sup></a> – A ledger instance can also be uniquely identified by its hash, which is a digital fingerprint of its contents.
-
 <a href="#footnote_from_2" id="footnote_2"><sup>2</sup></a> – Transactions with **tec** result codes are included in ledgers and not do the requested action. The rationale for this is that multiple transactions may be submitted in sequence, with the order of processing determined by a sequence number associated with an account (not to be confused with the ledger sequence number). To prevent a hol ...
```
Footnotes start from number 2.

ximinez reviewed (changes requested):
I'm not done yet, but I wanted to share some issues I've found so far.

mDuo13 committed:
Consensus diagram cleanup

- Remove unused HTML tx queue diagram and associated images
- Add white (not transparent) background to some diagrams
- Fix line breaks on a couple images
- Change a couple other concept pages to use the new diagrams
- Remove some newly unused old diagrams

mDuo13 committed:
Update tel codes tx queue link

ximinez commented on content/concepts/consensus-network/transaction-queue.md:
```diff
@@ -0,0 +1,69 @@
+
+1. **Consensus Round 1** - Each validator proposes a set of transactions to be included in the next ledger version. Each also keeps a queue of candidate transactions not currently proposed.
+
+2. **Consensus Round 2** - If a validator removes a transaction from its proposal in later rounds, it adds that transaction to its queue.
```
This is a little bit oversimplified, but probably appropriate for this level. At the code level, the transaction stays in the open/candidate ledger until consensus is finished, then any transactions in the open ledger that didn't get validated are reprocessed through the queue, potentially getting added if they don't pay enough.

ximinez commented on content/concepts/consensus-network/transaction-queue.md:
```diff
@@ -0,0 +1,69 @@
+
+5. **Building the Next Proposal** - Each validator prepares its proposal for the next ledger version, starting with queued transactions.
+
+6. **Adding to the Queue** - If the next proposed ledger is already full, incoming transactions are queued for a later ledger version.
```
This step leaves out a very important part of the design philosophy for fee escalation and the queue, which is that if a transaction is important enough, there is always a fee that a transaction can pay to bypass the queue and get directly into the open ledger. The ledger never really gets "full", but it gets to a point where transactions aren't cheap. I'm not sure how to better phrase this, but I'll give it a shot... (Feel free to rewrite as appropriate.)

> When the next proposed ledger contains enough transactions, incoming transactions that don't pay the escalated fee are queued for a later ledger version.

> If the next proposed ledger contains enough transactions, incoming transactions that pay enough to be valid, but not enough to meet the escalated cost may be queued for a later ledger version.

Ugh. This is why I'm a code monkey.

mDuo13 replied:
I left this sentence as-is, but added a parenthetical statement immediately after it:

> (Transactions that pay the [open ledger cost] can still get into the next proposed ledger even if it's "full", but the open ledger cost grows exponentially with each transaction added this way.)

ximinez commented on content/concepts/consensus-network/transaction-queue.md:
```diff
@@ -0,0 +1,69 @@
+
+[New in: rippled 0.33.0][]
+
+If a sending address has one or more transactions queued, that sender can "push" the existing queued transactions into the open ledger by submitting a new transaction with a high enough transaction cost to pay for all of them. Specifically, the new transaction must increase the total transaction cost of the queued transactions from the same sending address, including the new transaction, to cover ...
```
`Specifically, the new transaction must *pay* the total transaction cost of the queued transactions...`

mDuo13 replied:
I rewrote this paragraph a bit to improve even more on this point:

> Specifically, the new transaction must pay a high enough transaction cost to cover the [open ledger cost] of itself and each other transaction from the same sender before it in the queue. (Keep in mind that the open ledger cost increases exponentially each time a transaction pays it.) The transactions must still follow the other [queuing restrictions]...

The "of itself and each other transaction from the same sender before it in the queue" feels like quite a mouthful, but I haven't found a way to condense it any more than that.

ximinez commented on content/concepts/consensus-network/transaction-queue.md:
```diff
@@ -0,0 +1,69 @@
+
+* Wait for the queued transactions to be included in a validated ledger, _or_
+* Wait for the queued transactions to be permanently invalidated if the transactions have the [`LastLedgerSequence` field](reliable-transaction-submission.html#lastledgersequence) set, _or_
+* [Cancel the queued transactions](cancel-or-skip-a-transaction.html) by submitting a new transaction with the same sequence number.
```
`...and a higher cost.`

ximinez reviewed (changes requested):
I have several suggested changes for the Transaction Queue write up. Otherwise, it, and the diagram look really good.

> Reviewers: I'd appreciate your feedback on whether the unused diagram is useful scrap material or should be removed.
I am unsure of which files you are referring to that contain the unused diagram. In general, in code, it's better to remove than to comment out, especially if the unused stuff appears in source control history. So it can be found later if you need it. I think the same rule applies here.

gbarr01 commented on content/concepts/consensus-network/consensus.md:
```diff
@@ -6,109 +6,131 @@ This article provides a high level overview of the XRP Ledger, the information i
+- balances of XRP and [issued currencies](issued-currencies.html)
 - offers in the distributed exchange
 - network settings, such as [transaction costs](transaction-cost.html) and [reserve](reserves.html) amounts
 - a time stamp
```
```
< time stamp
---
> timestamp
```

gbarr01 commented on content/concepts/consensus-network/consensus.md:
```diff
@@ -6,109 +6,131 @@ This article provides a high level overview of the XRP Ledger, the information i
+[![Figure 5: Validators Propose and Revise Transaction Sets](img/consensus-rounds.png)](img/consensus-rounds.png)
 
-_Figure 5: Validators Propose Transaction Sets — At the start of consensus, nodes work with different sets of transactions. Rounds of proposals determine which transactions to apply to the ledger, and which must wait for a later round of consensus._
+_Figure 5: Validators Propose and Revise Transaction Sets — At the start of consensus, validators may have different sets of transactions. In later rounds, servers modify their proposals to match what their trusted validators proposed. This process determines which transactions the should apply to the ledger version currently being discussed, and which they should postpone for later ledger versio ...
```
```
< This process determines which transactions the should apply
---
> This process determines which transactions they should apply
```

gbarr01 commented on content/concepts/consensus-network/consensus.md:
```diff
@@ -6,109 +6,131 @@ This article provides a high level overview of the XRP Ledger, the information i
+
+[![Figure 8: Ledger is Validated When Supermajority of Peers Calculate the Same Result Result](img/consensus-declare-validation.png)](img/consensus-declare-validation.png)
+
+_Figure 8: Ledger is Validated When Supermajority of Peers Calculate the Same Result — Servers compare their calculated ledger with the hashes received from chosen validators. If not in agreement, the server must re-calculate or retrieve the correct ledger._
```
```
< Servers compare their calculated ledger with the hashes received from chosen validators. If not in agreement, the server must re-calculate or retrieve the correct ledger.
---
> Servers compare their calculated ledgers with the hashes received from their chosen validators. If not in agreement, the servers must recalculate or retrieve the correct ledger.

or

> Each server compares its calculated ledger with the hashes received from its chosen validators. If not in agreement, the server must recalculate or retrieve the correct ledger.
```

gbarr01 commented on content/concepts/consensus-network/consensus.md:
```diff
@@ -6,109 +6,131 @@ This article provides a high level overview of the XRP Ledger, the information i
 _Figure 4: Participants in the XRP Ledger Protocol_
 
-The nodes that receive, relay and process transactions may be either tracking nodes or validating nodes. Tracking nodes’ primary functions include distributing transactions from clients and responding to queries about the ledger. Validating nodes perform the same functions as tracking nodes and additionally contribute to advancing the ledger sequence <a href="#footnote_4" id="footnote_from_4"><su ...
+The servers that receive, relay and process transactions may be either tracking servers or validators. Tracking servers’ primary functions include distributing transactions from clients and responding to queries about the ledger. Validating servers perform the same functions as tracking servers and additionally contribute to advancing the ledger sequence <a href="#footnote_3" id="from_footnote_3" ...
```
Suggestion: 
```
< Tracking servers’ primary functions
---
> The primary functions of tracking servers
```

mDuo13 replied:
I didn't feel happy with multiple "primary" functions, so I changed this a little more:

> The major functions of tracking servers

gbarr01 commented on content/concepts/consensus-network/transaction-queue.md:
```diff
@@ -0,0 +1,69 @@
+* Transactions with an `AccountTxnID` field cannot be queued.
+* A single sending address can have at most 10 transactions queued at the same time. In order for a transaction to be queued, the sender must have enough XRP to pay all the XRP costs of all the sender's queued transactions including both the `Fee` fields and the sum of the XRP that each transaction could send. [New in: rippled 0.32.0][]
+* If a transaction affects how the sending address authorizes transactions, no other transactions from the same address can be queued behind it. [New in: rippled 0.32.0][]
+* If the transaction includes a `LastLedgerSequence` field, the value of that field must be at least **the current ledger index + 2**.
```
```
>
If a transaction affects
If the transaction includes
---
>
If a transaction affects
If a transaction includes
```

gbarr01 commented on content/concepts/consensus-network/transaction-queue.md:
```diff
@@ -0,0 +1,69 @@
+
+If a sending address has one or more transactions queued, that sender can "push" the existing queued transactions into the open ledger by submitting a new transaction with a high enough transaction cost to pay for all of them. Specifically, the new transaction must increase the total transaction cost of the queued transactions from the same sending address, including the new transaction, to cover ...
+
+This feature helps you work around a particular situation. If you submitted one or more transactions with a low cost, which got queued, you cannot send new transactions from the same address unless you do one of the following:
```
Suggestion: 
```
< If you submitted one or more transactions with a low cost, which got queued
---
> If you submitted one or more transactions with a low cost that were queued
```

gbarr01 commented:
@mDuo13 , some comments about the diagrams: 

**anatomy-of-a-ledger-complete.png**

1. "Count" in Total XRP Count should be lowercase
2. Put these on one line:
* Timestamp
* Total XRP Count
3. Move "State data" in a tad and put paren content on one line.
4. For Total XRP count, maybe use a graph that goes up?

**ledger-history.png**

> Each ledger version also contains a set of transactions and metadata about those transactions. The transactions it includes are only the ones that have been applied to the previous ledger version to create the new ledger version.

This (2nd) sentence is not reflected in the diagram. I see candidate ledgers with only new proposed transactions, and none that "have been applied to the previous ledger version". Do they get applied to the latest after the candidate has been validated? Doesn't seem so. What am I missing here?sed


**ledger-changes.png**

Should the arrow for "Hash identifies parent" be pointed in the other direction?


**consensus-rounds.png**

Where is this image? Did not see it in the list.


**consensus-calculate-validation.png**

In the 2nd section, shouldn't the green transactions be applied to the previous ledger (as the text in the diagram says)? ... shouldn't the arrow go in the opposite direction? You appear to be applying them to the new ledger.


**consensus-declare-validation.png**

```
< If ≥80% of a server's trusted validators got
---
> If ≥80% of a server's trusted validators get

< If a server got
---
> If a server gets
```

gbarr01 commented:
@mDuo13 , which is the the unused diagram? And should I review the other documents?

mDuo13 commented:
@gbarr01 I removed the unused diagram after thinking on it some more (I reached the same conclusion as @ximinez's recommendation above).

To answer your questions on ledger-history.png, I think you've misunderstood what it means to apply a transaction to a ledger. I'm not sure what I can clarify in the document to make this make more sense. What I can say is: in the consensus process, "Ledger version" objects are immutable. (The server does keep a mutable "current open ledger" as a tentative working copy , but the consensus process doesn't actually use the open ledger at all.) When you apply transactions to any ledger version, the result is a new ledger version whose parent is the one you used as a base. The "candidate ledger versions" you see in the diagram are possible results of applying (slightly different sets of) proposed transactions to previous ledger versions. To reiterate, a ledger version contains:

- Metadata about itself, including an identifying hash of its contents and the identifying hash of its direct predecessor ("parent ledger")
- The latest state data
- Transactions that have been newly applied since the parent ledger version
    - This also includes metadata about exactly what happened when applying each transaction (essentially, each transaction's metadata shows a "diff" of the state data as a result of processing the transaction)

mDuo13 committed:
Consensus/TxQ text cleanup

mDuo13 commented:
66f7374 contains the changes to the Markdown files. The fixes to diagrams are coming next (after I get my image editor setup in place again). Notes on the specific diagrams:

- `consensus-rounds.png` is not in the diff because it's unchanged from PR #445
- `anatomy-of-a-ledger-complete.png` - I have the fixes for this ready to export. It originally looked more like what you described, but I think screwing with the aspect ratio in Google Slides broke it.
    - I won't be changing the XRP count to a graph going up. It's a graph going down because the total supply of XRP can only decrease as XRP is destroyed.
- `ledger-changes.png` "hash identifies parent" direction—I think it sort of makes sense either way, but your suggestion is better. Changing it.
- `consensus-calculate-validation.png` - As clarified earlier, old ledgers are used as a basis to apply new transactions. The diagram is correct.
- `consensus-declare-validation.png` - I think this is one case where past tense is appropriate and important, because we are describing a state in which servers are looking back on and evaluating results that took place in the recent past from the perspective of the moment in question. [The fact that it works this way is a relevant part of how the consensus algorithm approaches an unsolvable problem.](https://developers.ripple.com/consensus-principles-and-rules.html#consensus-can-fail)

mDuo13 committed:
Consensus diagrams: tweaks per @gbarr01 review
